### PR TITLE
Major UI rework to fix multiple UI/UX bugs

### LINKED ISSUE #151: [PE-D] No command to switch between ingredient and recipe list
Personally I like how clean your UI is, however I noticed that there is currently no command to switch the current displayed list from either the ingredients list to the recipe list or vice versa. As the programme should be optimised for a fast typer, I feel like there should be a command to allow for the user to switch between list as otherwise it breaks this requirements of optimisation for a fast typer.

-------------
Labels: `type.FeatureFlaw` `severity.Medium`
original: KT27Learn/ped#3

### LINKED ISSUE #164: [PE-D] Switch Tabs does not reset ingredient view
When i switch between the ingredient and recipe list, the selected ingredient/receipe remains the same. This prevents me from clicking the same ingredient again to display their information. I am not sure whether this is the intended implementation so Im raising this bug for you to take a look.

[Screen Recording 2021-10-29 at 4.45.41 PM.mov.zip](https://raw.githubusercontent.com/KT27Learn/ped/main/files/0824f212-dc2f-4a56-86af-8006f4e58489.zip)

-------------
Labels: `severity.Low` `type.FunctionalityBug`
original: KT27Learn/ped#5

### LINKED ISSUE #177: [PE-D] GUI does not respond fully to list ingredient and list recipe
When previously on the recipe page and I want to list all the ingredient instead, the `list ingredient` command does not change the page: the menu still shows all the recipes instead. Should update the GUI to show all the ingredient instead. Same case for on the ingredient page and want to list all the recipes.

-------------
Labels: `severity.Medium` `type.FeatureFlaw`
original: LiuZhi1an/ped#7

### LINKED ISSUE #181: [PE-D] GUI does not respond immediately after executing the cook recipe command
After executing the cook recipe command, there is no more potato and the status of potato should be missing and the color of the potato in the recipe list does not turn red. It will only do so when I click on another recipe.

-------------
Labels: `severity.Medium` `type.FeatureFlaw`
original: LiuZhi1an/ped#8

### LINKED ISSUE #190: Cook recipe does not update recipe ingredient tags
When you cook recipe and use up ingredients, the tags on recipe are not updated to reflect if there's enough ingredient. 

For example, if there are 2 eggs in ingredients, and recipe uses 2 eggs, after cook recipe, the recipe egg tag will still be blue instead of expected red.

@NayLin-H99

### LINKED ISSUE #197: [PE-D] View of a recipe is cleared when user edits the recipe
The view of the recipe should be kept if the same recipe is edited

Before:

After:

This has a different behavior if a different recipe is viewed when editing

Before:

After:

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: SpdPnd98/ped#9

### PR DESCRIPTION
**Pull request by rootkie**
This change should fix #197 fix #181 fix #177 fix #164 fix #151 fix #190

## Changed Behaviours

1. tab switching is implemented for **View List Find Edit Add** command. Not sure if **Delete** should switch tab, because it might be confusing if the tab switches and the item is already gone.
2. tab switching will remember the current active item, only active item will be selected
3. Edits / Cook Recipe will update both tab list and the active window recipe

All tab swiitch related behaviours will work through both GUI and CLI command.

## Implementation details

I have grouped the UI state under UiState class under Ui package. It will serve as an observable for commands executed to push notification / update GUI components.

GUI updates are handled through the use of Events. There are 2 kind of Events defined, ActiveItemChangeEvent and TabSwitchEvent.

ActiveItemChangeEvent is handled by both TabList and ActiveItemPanel. TabList removes all current selection then select the ActiveItem sent through the event. ActiveItemPanel simply updates the currently displayed item.

TabSwitchEvent simply switch the tab that is active.

GUI generated Events will back propagate the changes made to UI state through callbacks. You can see it in both IngredientListPanel and RecipeListPanel.

## Testing

I have attached the Jar file based off of this commit for easy testing of new behaviours. Let me know if there are any behaviour that you think doesn't make sense or is buggy.

[addressbook.zip](https://github.com/AY2122S1-CS2103T-W11-1/tp/files/7461543/addressbook.zip)

### CODE REVIEW

**Review by pattan9955 — APPROVED** [review]
Tested and works fine, LGTM! 🐒

File: src/main/java/fridgy/ui/MainWindow.java [hunk 1]
```diff
@@ -118,20 +118,40 @@ private void setAccelerator(MenuItem menuItem, KeyCombination keyCombination) {
...
+
+
+        // Initialise event listeners.
+        this.getRoot().addEventFilter(ActiveItemChangeEvent.RECIPE, activeItemPanel::handleRecipeEvent);
+        this.getRoot().addEventFilter(ActiveItemChangeEvent.INGREDIENT, activeItemPanel::handleIngredientEvent);
+        this.getRoot().addEventFilter(ActiveItemChangeEvent.CLEAR, activeItemPanel::handleClearEvent);
+
+        this.getRoot().addEventFilter(TabSwitchEvent.CHANGE, tabListPanel::handleTabChange);
+        this.getRoot().addEventFilter(ActiveItemChangeEvent.RECIPE, tabListPanel::handleActiveRecipe);
+        this.getRoot().addEventFilter(ActiveItemChangeEvent.INGREDIENT, tabListPanel::handleActiveIngredient);
```
Comment (NayLin-H99) [inline src/main/java/fridgy/ui/MainWindow.java:142]: Very good use of event handlers! 🐒

**Review by NayLin-H99 — CHANGES_REQUESTED** [review]
Excellent use of events and event handlers. 

I think you might want to fire another event to clear the ActiveItemPanel when the current item viewed is being deleted. 🐒  I am not sure if you need to switch the tab, but I think you can just do it for consistency sake. 

Other than that LGTM! 🙈

File: src/main/java/fridgy/ui/MainWindow.java (hunk 1 above)
Comment (rootkie) [reply src/main/java/fridgy/ui/MainWindow.java:142]: Ok, changing delete command to switch tab and clear window if deleted item is active.

**Review by NayLin-H99 — APPROVED** [review]
LGTM! 🐒

**Pull request merged**